### PR TITLE
Enhance Find command to use for multiple fields at once.

### LINKED ISSUE #109: Have the ability to find contacts based on fields other than name.

### PR DESCRIPTION
**Pull request by igezt**
Closes #109 

Usage e.g.:
`find n/Pierce n/Stanley mt/CS2103T`
will find all contacts that have either Pierce **or** Stanley in their name and has CS2103T in their modules taken.

Note: Its an either or for the name field because there are multiple names specified.

Other changes:
Renamed `EditPersonDescriptor -> PersonDescriptor` and `EditPersonCommandParser -> PersonCommandParser`.
Added more tests for `FindCommandParser` as well.

> **ONLY MERGE AFTER FACULTY PR HAS BEEN MERGED**

### CODE REVIEW

File: src/main/java/seedu/address/model/person/fields/Address.java [hunk 1]
```diff
@@ -42,6 +42,13 @@ public static boolean isValidAddress(String test) {
         return test.matches(VALIDATION_REGEX);
     }
 
+    /**
+     * Returns true if test string is contained within the value of the field.
+     */
+    public boolean contains(String test) {
+        return this.value.toUpperCase().contains(test.toUpperCase());
+    }
```
Comment (hansstanley) [inline src/main/java/seedu/address/model/person/fields/Address.java:50]: As mentioned in `Person.java`, this method can be abstracted into a `Field` abstract class or interface (same applies for the rest of the fields).

> although if that's too much work I can understand up to your own discretion there's so much work this week and so little time aaaaahhhhhh

File: src/main/java/seedu/address/storage/user/JsonAdaptedUser.java [hunk 2]
```diff
@@ -42,15 +42,15 @@ public class JsonAdaptedUser extends JsonAdaptedPerson {
...
-                           @JsonProperty("events") List<JsonAdaptedEvent> events) {
+                            @JsonProperty("email") String email, @JsonProperty("address") String address,
+                            @JsonProperty("race") String race,
+                            @JsonProperty("major") String major, @JsonProperty("gender") String gender,
+                            @JsonProperty("comms") String comms,
+                            @JsonProperty("modules") List<JsonAdaptedNusMod> modules,
+                            @JsonProperty("tagged") List<JsonAdaptedTag> tagged,
+                            @JsonProperty("favorite") String isFavorite,
+                            @JsonProperty("faculty") String faculty,
+                            @JsonProperty("events") List<JsonAdaptedEvent> events) {
```
Comment (hansstanley) [inline src/main/java/seedu/address/storage/user/JsonAdaptedUser.java:53]: Good catch 👍🏼

**Review by hansstanley — CHANGES_REQUESTED** [review]
Some suggestions to reduce code duplication among other benefits, but the changes compared to what it is now could be quite significant, so up to your discretion.

Overall good effort 👍🏼

File: src/main/java/seedu/address/logic/commands/FindCommand.java [hunk 3]
```diff
@@ -14,14 +14,15 @@ public class FindCommand extends Command {
 
     public static final String COMMAND_WORD = "find";
 
-    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Finds all persons whose names contain any of "
-            + "the specified keywords (case-insensitive) and displays them as a list with index numbers.\n"
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Finds all persons who contains any of "
+            + "the specified keywords (case-insensitive) in the specified fields and displays them as a list \n"
+            + "with index numbers.\n"
```
[the remark this answers is not in the document]
Comment (igezt) [reply src/main/java/seedu/address/logic/commands/FindCommand.java:19]: HAHAHA alright thank you

File: src/main/java/seedu/address/model/person/PersonContainsKeywordsPredicate.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    /**
+     * Stores the name in the predicate.
+     */
+    public PersonContainsKeywordsPredicate withName(Collection<String> name) {
+        if (!this.keywords.containsKey(NAMEKEY)) {
+            this.keywords.put(NAMEKEY, new HashSet<>());
+        }
+        this.keywords.get(NAMEKEY).addAll(name);
+        return this;
+    }
```
[the remark this answers is not in the document]
Comment (igezt) [reply src/main/java/seedu/address/model/person/PersonContainsKeywordsPredicate.java:43]: OOOO GOOD SUGGESTION. THANKS FAM

File: src/main/java/seedu/address/model/person/fields/Field.java [hunk 5]
```diff
@@ -20,8 +22,16 @@ public Field(String value) {
...
+        for (String t: test) {
+            if (t.isEmpty()) {
+                continue;
+            }
+            if (this.value.toUpperCase().contains(t.toUpperCase())) {
+                return true;
+            }
+        }
+        return false;
     }
```
Comment (hansstanley) [inline src/main/java/seedu/address/model/person/fields/Field.java:35]: With reference to the previous comment, this has an OR relationship!

File: src/main/java/seedu/address/model/person/Person.java [hunk 6]
```diff
@@ -209,6 +212,50 @@ public Person unfavorite() {
                 major, modules, race, tags, comms, newFavorite, faculty);
     }
 
+    /**
+     * Returns true if the person contains one of the keywords in all the fields specified, returns false otherwise.
+     */
+    public boolean contains(HashMap<PredicateKey, Set<String>> keywords) {
```
[the remark this answers is not in the document]
Comment (hansstanley) [reply src/main/java/seedu/address/model/person/Person.java:218]: Also might have to consider this when writing tests

**Review by hansstanley — APPROVED** [review]
Added some comments on search behaviour that I think are important to consider, otherwise approved so you can merge after.

**Pull request merged**